lib/executors.go

```diff
@@ -0,0 +1,341 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2019 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package lib
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/loadimpact/k6/stats"
+	"github.com/loadimpact/k6/ui/pb"
+	"github.com/sirupsen/logrus"
+	null "gopkg.in/guregu/null.v3"
+)
+
+//TODO: remove globals and use some type of explicit dependency injection?
+//nolint:gochecknoglobals
+var (
+	executorConfigTypesMutex   sync.RWMutex
+	executorConfigConstructors = make(map[string]ExecutorConfigConstructor)
+)
+
+// ExecutionStep is used by different executors to specify the planned number
+// of VUs they will need at a particular time. The times are relative to their
+// StartTime, i.e. they don't take into account the specific starting time of
+// the executor, as that will be considered by the external execution executor
+// separately.
+//
+// A slice [{t1, v1}, {t2, v2}, {t3, v3}, ..., {tn, vn}] of execution steps
+// means that a executor will need 0 VUs until t1, it will need v1 number of
+// VUs from time t1 until t2, need v2 number of VUs from time t2 to t3, and so
+// on. t1 is usually 0, tn is usually the same as GetMaxDuration() and vn is
+// usually 0.
+//
+// Keep in mind that t(i) may be exactly equal to t(i+i), when there's an abrupt
+// transition in the number of VUs required by a executor. For example, the
+// variable-looping-vus executor may have 0-duration stages, or it may scale up
+// VUs in its last stage right until the end. These
+//
+// []ExecutionStep is also used by the ExecutorConfigMap, to represent the
+// amount of needed VUs among all executors, during the whole execution of a
+// test script. In that context, each executor's StartTime is accounted for and
+// included in the offsets.
+type ExecutionStep struct {
+	TimeOffset      time.Duration
+	PlannedVUs      uint64
+	MaxUnplannedVUs uint64
+}
+
+//TODO: make []ExecutionStep or []ExecutorConfig their own type?
+
+// ExecutorConfig is an interface that should be implemented by all executor config types
```

Comment:
Can you expand on this comment? It sounds a bit obvious and self-referential right now, but it doesn't explain why a separate configuration interface specifically for executors is needed.

Could parts of `ExecutorConfig` be handled by a new config API? I'm not suggesting we add that as well to this PR :smile:, but maybe we could structure this better to allow for that in the near future.

For example, methods like `Validate()` or ones that just return configuration values like `GetEnv()` and `GetExec()` look like they could be part of a generic `config` package. And others like `GetName()`, `GetType()` and `IsDistributable()` could become parts of `Executor` itself. After all, they all return information about an executor, not an executor _configuration_, so that abstraction doesn't make much sense to me.

Essentially, I'm suggesting decoupling of executors and configuration by keeping configuration-related methods in the current system or slightly adapting it for a better fit, moving others to `Executor`, and dropping this new interface.

Let me know if this makes sense, or if it would be feasible at this point.

Reply:
The biggest reason is `ExecutorConfigMap`, i.e. the `map[string]ExecutorConfig` type we use to parse and validate the `execution` option. Having this interface allows us to automagically unmarshal the very different configs in the single data structure, and then validate them and treat them together in the rest of the codebase, not caring about their specific types one jot.

> After all, they all return information about an executor, not an executor configuration, so that abstraction doesn't make much sense to me.

In my mind, most of the methods in this interface are about querying information about an executor config - after all, that's the things users will specify in the `execution` options.

The only exception, `IsDistributable()`, is about querying an inherent property of the executor type. While it would make some sense to be an actual `Executor` method, it's not actually a property of its instances, rather it's a property of its *type*, i.e. it would have been a static method in other languages. But since go doesn't have that, and since we need to know whether a particular executor is distributable **before** we initialize it (for `k6 cloud` where we never actually initialize executors locally and for the future native k6 distributed execution), I added it to the `ExecutorConfig` type.

Reply:
> The biggest reason is ExecutorConfigMap, i.e. the map[string]ExecutorConfig type we use to parse and validate the execution option. Having this interface allows us to automagically unmarshal the very different configs in the single data structure, and then validate them and treat them together in the rest of the codebase, not caring about their specific types one jot.

That is a compelling use case, and I see why you did it that way.

I still think configuration should be "dumber" (set/get only), and not tied to a particular app or functionality.


> In my mind, most of the methods in this interface are about querying information about an executor config - after all, that's the things users will specify in the execution options.

True, you can look at it that way. For me configuration is something best handled once up front / on startup in a generic way, and then passed on to (or read by) different components that need it. It shouldn't be critical to the application besides being an in-memory key-value store.

Which is why I see `NewExecutor()` and `GetExecutionRequirements()` as being outside the responsibility of a configuration type.

In my ideal design, the pseudo-lifecycle would be something like:

1. `k6/cmd` gathers config sources, assembles a `k6.config.Config` object, and validates it.
2. `k6/engine` or `k6/execution` receives a validated `Config`, walks through `Config.Execution` and instantiates the required executors into an `ExecutorMap` or similar.
3. `ExecutorMap` is used the same way as `ExecutorConfigMap` is currently, without configuration being a runtime requirement. So all current `GetName()`, `GetStartTime()`, etc. methods will be retrieved from the already configured executor instance itself.

This would require an overhaul of the current config system, but if we restructured only what's required for the new executors in this way, it could serve as a blueprint for the config refactor we could expand on later.

> since we need to know whether a particular executor is distributable before we initialize it

I see, and agree that it would make more sense for this to be a static method, but I would still prefer having an `instance.isDistributable()` check in the constructor, than handling it with a specific configuration type.


Not sure, this would be a lot of work to refactor currently, but it would bother me if we have a clean `k6/config` package one day that parts of the configuration is being validated and used as a functional requirement in other parts of the codebase.

Reply:
1. and 2. is pretty much what happens right now, only instead of using a type switch to match the config type to the executor type, we "upgrade" the config to a fully fledged instance. And the only ways I see we can do it with Go's type system are: the current one, a huge type `switch`, and reflection. 

Point 3. doesn't work, because in k6 we have situations in which we need to know things about the executors **without** initializing them. `isDistributable` is one of them, but the names, start times and even the execution plan may be needed for things like `k6 inspect`, `k6 cloud`, `k6 archive`, etc.

> but it would bother me if we have a clean `k6/config` package one day that parts of the configuration is being validated and used as a functional requirement in other parts of the codebase.

I'm not sure in this situation it makes sense to have an executor configuration loosely coupled with its actual  implementation. That is, I don't see any benefits to having the `execution` configuration live its own separate life in another package with other configuration, and the executor implementations (i.e. the `Run()` methods and helpers) living their own life at some other place in the code. The current format seem much easier to read and reason about than the alternative - everything regarding a single executor (configuration, validation, execution) is well encapsulated and easy to reason about.

Reply:
I agree with your last point, the current design solves things nicely for executors. My main concern is that configuration is an intrinsic part of it, whereas configuration to me should be dumb/static and completely separate from app functionality. I would like to push towards a more unified configuration approach (getting rid of `lib.RuntimeOptions`, cleaning up how overriding works, etc.), but a new executor-specific config would fragment this further and make the unification more difficult.

That said, I trust your judgment, so if you don't see a problem with this, we can leave it as is. :)

Reply:
> My main concern is that configuration is an intrinsic part of it, whereas configuration to me should be dumb/static and completely separate from app functionality

This should probably be a wider discussion, but I'm not sure I agree with you on this. In a dynamically typed language like JS/PHP/Python/etc., or in a statically typed language with a more flexible type system (Java, Rust, etc.), it might be the better approach, but I don't think we'll benefit from it in Go.

For one, I think the configuration should be at least smart enough that it can validate itself - putting the validation in the config layer instead of the execution later just makes error handling so much easier and cleaner. Then the executors don't have to bother with any validation - they "know" that if they were initialized, they have a valid configuration. And having a separate config data structure and a central validation layer also doesn't seem like a good idea - you'd loose locality (i.e. would be harder to reason about and maintain) and would end up with huuge validation routines (sort of like the current `.Apply()` mess).

I realize that "spawning" an executor instance from a config one is strange. I don't like it very much either, but unfortunately that's the only type-safe way I could think to do it. Any other implementation that I can think of would require *runtime* type assertions or reflection, whereas with the current approach we have compile-time safety guarantees. And if we get around the slight strangeness, I honestly find it pretty easy and efficient to work with: https://github.com/loadimpact/k6/blob/edd03a37360c7efbe05fe98f835050016fb7f06e/core/local/local.go#L70-L71

Besides the slight strangeness of the constructors, I struggle to see how coupling the executor configs with the executor implementations that depend on them is actually problematic. I read through https://github.com/loadimpact/k6/issues/883 and don't think anything there besides the serialization issues of null-able types applies in this case, and that one is a separate issue. I might be biased, having written the code, but I don't see how this config paradigm would frustrate anything we'd like to do in the future.

Reply:
@na-- Fair enough, thanks for explaining. I'll fallback on your judgment here, since I'm not that familiar with the codebase, or Go for that matter.

I have some ideas about configuration we could discuss as part of #883, but I'll resolve this thread for now.